lib/libm/man/lrint.3
.\" 	$OpenBSD: src/lib/libm/man/lrint.3,v 1.3 2011/07/07 00:54:16 martynas Exp $
.\" $NetBSD: lrint.3,v 1.1 2005/09/16 15:26:47 wiz Exp $
.\"
.\" Copyright (c) 2005 David Schultz <das@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: /repoman/r/ncvs/src/lib/msun/man/lrint.3,v 1.2.2.2 2005/03/01 16:18:39 brueffer Exp $
.\"
.Dd $Mdocdate: May 31 2007 $
.Dt LRINT 3
.Os
.Sh NAME
.Nm llrint ,
.Nm llrintf ,
.Nm llrintl ,
.Nm lrint ,
.Nm lrintf ,
.Nm lrintl
.Nd convert to integer
.Sh SYNOPSIS
.In math.h
.Ft long long
.Fn llrint "double x"
.Ft long long
.Fn llrintf "float x"
.Ft long long
.Fn llrintl "long double x"
.Ft long
.Fn lrint "double x"
.Ft long
.Fn lrintf "float x"
.Ft long
.Fn lrintl "long double x"
.Sh DESCRIPTION
The
.Fn lrint
function returns the integer nearest to its argument
.Fa x
according to the current rounding mode.
.Pp
The
.Fn llrint ,
.Fn llrintf ,
.Fn llrintl ,
.Fn lrintf ,
and
.Fn lrintl
functions differ from
.Fn lrint
only in their input and output types.
.Sh RETURN VALUES
The
.Nm llrint ,
.Nm llrintf ,
.Nm llrintl ,
.Nm lrint ,
.Nm lrintf ,
and
.Nm lrintl
functions return the integer nearest to their argument
.Fa x
according to the current rounding mode.
If the rounded result is too large to be represented as a
.Vt long long
or
.Vt long
value, respectively,
.\" an invalid exception is raised and
the return value is undefined.
.\" Otherwise, if
.\" .Fa x
.\" is not an integer,
.\" .Fn lrint
.\" raises an inexact exception.
.\" If
.\" .Fa x
.\" is too large, a range error may occur.
.Sh SEE ALSO
.Xr lround 3 ,
.Xr math 3 ,
.Xr rint 3
.Sh STANDARDS
The
.Fn llrint ,
.Fn llrintf ,
.Fn llrintl ,
.Fn lrint ,
.Fn lrintf ,
and
.Fn lrintl
functions conform to
.St -isoC-99 .


lib/libm/src/lrint.c
/*	$OpenBSD: src/lib/libm/src/Attic/lrint.c,v 1.1 2005/11/17 20:07:40 otto Exp $	*/
/* $NetBSD: lrint.c,v 1.3 2004/10/13 15:18:32 drochner Exp $ */

/*-
 * Copyright (c) 2004
 *	Matthias Drochner. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/limits.h>
#include <math.h>
#include <ieeefp.h>
#include <machine/ieee.h>
#include "math_private.h"

#ifndef LRINTNAME
#define LRINTNAME lrint
#define RESTYPE long int
#define RESTYPE_MIN LONG_MIN
#define RESTYPE_MAX LONG_MAX
#endif

#define RESTYPE_BITS (sizeof(RESTYPE) * 8)

static const double
TWO52[2]={
  4.50359962737049600000e+15, /* 0x43300000, 0x00000000 */
 -4.50359962737049600000e+15, /* 0xC3300000, 0x00000000 */
};

RESTYPE
LRINTNAME(double x)
{
	u_int32_t i0, i1;
	int e, s, shift;
	RESTYPE res;

	GET_HIGH_WORD(i0, x);
	e = i0 >> 20;
	s = e >> DBL_EXPBITS;
	e = (e & 0x7ff) - DBL_EXP_BIAS;

	/* 1.0 x 2^-1 is the smallest number which can be rounded to 1 */
	if (e < -1)
		return (0);
	/* 1.0 x 2^31 (or 2^63) is already too large */
	if (e >= (int)RESTYPE_BITS - 1)
		return (s ? RESTYPE_MIN : RESTYPE_MAX); /* ??? unspecified */

	/* >= 2^52 is already an exact integer */
	if (e < DBL_FRACBITS) {
		/* round, using current direction */
		x += TWO52[s];
		x -= TWO52[s];
	}

	EXTRACT_WORDS(i0, i1, x);
	e = ((i0 >> 20) & 0x7ff) - DBL_EXP_BIAS;
	i0 &= 0xfffff;
	i0 |= (1 << 20);

	shift = e - DBL_FRACBITS;
	if (shift >=0)
		res = (shift < 32 ? (RESTYPE)i1 << shift : 0);
	else
		res = (shift > -32 ? i1 >> -shift : 0);
	shift += 32;
	if (shift >=0)
		res |= (shift < 32 ? (RESTYPE)i0 << shift : 0);
	else
		res |= (shift > -32 ? i0 >> -shift : 0);

	return (s ? -res : res);
}
